linux-v4l2: Expanded error and debug output and added automatic device reset feature

esden opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description

This pull request adds the following debug output improvements:
* Timeout errors are now LOG_ERROR instead of LOG_DEBUG. This made it hard to track down video device timeout issues.
* Additional diagnostic debug output, showing frame buffer states.
* All video data stream outputs are now prefixed with the name of the device for easier readability.
* Added DEBUG output that prints all buffer states when a timeout occurs.

Additionally, this pull request adds a setting to the linux-v4l2 plugin to enable automatic device resets. (Not enabled by default)

When the `select` times out (no frames available from the video capture device) for longer than the set amount of "frame times" (minimum 2, maximum 120) the plugin disable `streaming` and re-enable it again. Effectively this resets the video capture device.

### Motivation and Context

This pull request is the result of many days of work trying to debug issues with my USB, HDMI video capture devices. 

It improves the error output. Without the patch OBS was completely silent when the HDMI capture devices were freezing. Now it will indicate that the video capture devices are timing out and print which device it is. The added debug output is even more useful when compiled with `-DCMAKE_BUILD_TYPE=Debug`.

The automatic reset feature, turns out to be very useful with devices like the Elgato Cam Link 4k in bulk mode on some USB chip sets, as well as Magewell USB Capture HDMI in general. Those like to "lock up" or "freeze". 

Note: (For those that randomly find this Pull Request.) A huge stability improvement for Elgato Cam Link 4k on Linux is switching it to Isochronous mode using the Windows tool.

I left the automatic resets off by default because it might cause issues, in cases where the device is timing out a lot, but still work. (aka. dropping a lot of frames) In other cases where the devices tend to freeze, the automatic reset makes a completely unusable streaming setup, usable again with just a few dropped frames.

The timeout is calculated in the terms of frame time multiples. The reason is, that a "reasonable" timeout (not too long and not too short), is tightly coupled to the framerate of the video captured. The "frame time" multiplier can be set per device depending on how reliable they are. Experiments have shown that 5x frame times is a reasonable compromise between reaction time and false positive resets in most situations. (that is why the default is 5x frame times)

### How Has This Been Tested?
Many hours of video recording and streaming in multiple configurations.

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
- New feature (non-breaking change which adds functionality)
- Performance enhancement (non-breaking change which improves efficiency)
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [ ] I have included updates to all appropriate documentation.

WizardCM commented:
I feel the logging changes should be in a separate PR from the automatic device reset.

esden commented:
@WizardCM I was thinking about splitting it. The problem is that both topics touch essentially the same lines of code. It would be pretty messy to split them apart. And they both enhance each other regarding testing of the pull request.

But if you insist, I can try to split it. That said, after spending a significant time on this yak shave. I don't know when I will have time to do that. :)

RytoEX commented on plugins/linux-v4l2/data/locale/en-GB.ini:
```diff
@@ -13,4 +13,5 @@ ColorRange.Default="Default"
 ColorRange.Full="Full"
 CameraCtrls="Camera Controls"
-
+AutoresetOnTimeout="Reset On Timeout"
+FramesUntilTimeout="Frames Until Timeout"
```
Do not include changes to locale files other than en-US.  All translation work is done via Crowdin, as noted in [our Contributing Guidelines](https://github.com/obsproject/obs-studio/blob/4a890e1da9a27003c796527b4a603312514e9eac/CONTRIBUTING.rst#quick-links-for-contributing).

RytoEX commented on plugins/linux-v4l2/v4l2-input.c:
```diff
@@ -14,6 +14,7 @@ GNU General Public License for more details.
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
+
```
Avoid unnecessary whitespace changes.

RytoEX reviewed (changes requested):
Per [our commit guidelines](https://github.com/obsproject/obs-studio/blob/4a890e1da9a27003c796527b4a603312514e9eac/CONTRIBUTING.rst#commit-guidelines):
1. Do not end commit message subjects with a period.
2. Use a present tense verb for the first word after the commit prefix.
3. Try to keep commit message subjects, not including the prefix, at 50 characters or less.

Additionally, do not include a separate commit for style fixes to your commits.

kkartaltepe commented:
Changes should really be squashed at least (and preferably split as mentioned).

esden commented:
All right, thanks for the input. I will open two new pull requests with the log output changes and the auto reset feature. Squashed, cleaned and adjusted as requested.

esden commented:
For reference: this pull request is replaced by #4004 and #4005

kkartaltepe commented:
For the future, there was no need to close this PR. You could have force pushed and adjust the name to reflect the part you left here.

esden commented:
@kkartaltepe all right. Will do that in the future.
